File: Difficulty: Medium/Search in Rotated Sorted Array/search-in-rotated-sorted-array.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

public class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);

        int t = Integer.parseInt(in.readLine().trim());
        while (t-- > 0) {
            String line = in.readLine();
            String[] tokens = line.split(" ");

            // Create an ArrayList to store the integers
            ArrayList<Integer> array = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : tokens) {
                array.add(Integer.parseInt(token));
            }

            int[] arr = new int[array.size()];
            int idx = 0;
            for (int i : array) arr[idx++] = i;

            int key = Integer.parseInt(in.readLine().trim());

            out.println(new Solution().search(arr, key));
        }
        out.close();
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {
    int search(int[] arr, int key) {
        // Complete this function
        int l=0;
        int r=arr.length-1;
        while(l<=r)
        {
            int m=(l+r)/2;
            if(arr[m]==key)
            {
                return m;
            }
            if(arr[l]<=arr[m])
            {
                if(arr[l]<=key && key <=arr[m])
                {
                    r=m-1;
                }
                else
                {
                    l=m+1;
                }
            }
            else
            {
                if(arr[m]<=key && key<=arr[r])
                {
                    l=m+1;
                }
                else
                {
                    r=m-1;
                }
            }
        }
        return -1;
    }
}

File: Difficulty: Medium/Search in Rotated Sorted Array/search-in-rotated-sorted-array.py
#User function Template for python3

class Solution:
    def search(self,arr,key):
        # Complete this function
        i = 0
        n = len(arr)
        j = n - 1

        while i <= j:
            mid = (i + j) // 2
            if arr[mid] == key:
                return mid
            if arr[mid] < arr[j]:
                if arr[mid] <= key <= arr[j]:
                    i = mid + 1
                else:
                    j = mid - 1
            else:
                if arr[i] <= key <= arr[mid]:
                    j = mid - 1
                else:
                    i = mid + 1
        return -1

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())

    for _ in range(t):
        A = list(map(int, input().strip().split()))
        k = int(input())
        ob = Solution()
        print(ob.search(A, k))
        print("~")

# } Driver Code Ends